### ISSUE TITLE
01-logistic-regression.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/machine-learning-elius123ef#2]
Hola @elius123ef aunque hay fuentes que hacen el escalado y el encoding antes en este pedíamos que se hiciera para no tener data leakage con el orden de 

- split
- encoding
- escalado

¡Te paso las issues!: 

### **Error: Data Leakage por escalado antes del split**

**Problema principal:**

Escalas las variables **antes** de hacer el split train/test:

```python
# INCORRECTO
scaler = StandardScaler()
df_scaled = pd.DataFrame(scaler.fit_transform(df_copy[num_vars + ['contact_intensity']]),
                         columns=num_vars + ['contact_intensity'])

# Concatenas todo
X_final = pd.concat([df_scaled, df_encoded, df_job_edu, df_copy[['is_retired_no_loan', 'is_single_high_edu']]], axis=1)

# Split DESPUÉS del escalado
X_train, X_test, y_train, y_test = train_test_split(
    X_final, y_final, test_size=0.3, random_state=42, stratify=y_final)
```

**Consecuencias:**
- El `StandardScaler` calcula media y desviación con **todo el dataset** (incluyendo test)
- El modelo "ve" información del conjunto de test durante el entrenamiento
- Las métricas reportadas (accuracy 80.93%, recall 76.65%) están infladas
- En producción el modelo rendirá peor

---

#### 2. **Factorización antes del split**

```python
# NCORRECTO
df_copy['job'] = pd.factorize(df['job'])[0]
df_copy['marital'] = pd.factorize(df['marital'])[0]
# ... etc
```

Aunque `factorize()` es menos crítico que el escalado, sigue siendo data leakage porque asigna códigos basándose en todo el dataset.

#### 3. **OneHotEncoder aplicado antes del split**
```python
df_encoded = pd.get_dummies(df_copy[cat_vars], drop_first=True)
```

`get_dummies()` crea columnas basándose en todas las categorías del dataset completo, no solo del train.

#### 4. **SelectKBest aplicado en el dataset completo**
```python
# INCORRECTO
selector = SelectKBest(score_func=f_classif, k=20)
X_selected = selector.fit_transform(X_clean, y_final)  # Usas TODO el dataset
```

La selección de características debe hacerse **solo con los datos de entrenamiento**.

---

### **Aspectos positivos:**

1. **Excelente EDA:** análisis univariante, bivariante y multivariante muy completo
2. **Ingeniería de características creativa:**
   - Variables binarias: `is_retired_no_loan`, `is_single_high_edu`
   - Variable derivada: `contact_intensity`
   - Binning de duration: `duration_bin`
   - Interacciones: `job_edu`
3. **Análisis de multicolinealidad (VIF):** muy profesional
4. **Uso de `class_weight='balanced'`:** aborda el desbalanceo de clases
5. **Comparación de modelos:** contrasta resultados con/sin balanceo
6. **Stratify en split:** mantiene proporciones de clases

---
